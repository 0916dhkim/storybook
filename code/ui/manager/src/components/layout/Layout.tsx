import React, { useEffect, useLayoutEffect, useState } from 'react';
import { styled } from '@storybook/theming';
import type { Props, LayoutState, ManagerLayoutState } from './Layout.types';
import { useDragging } from './useDragging';

const MINIMUM_CONTENT_WIDTH_PX = 300;

const layoutStateIsEqual = (state: ManagerLayoutState, other: ManagerLayoutState) =>
  state.navSize === other.navSize &&
  state.bottomPanelHeight === other.bottomPanelHeight &&
  state.rightPanelWidth === other.rightPanelWidth &&
  state.panelPosition === other.panelPosition;

/**
 * Manages the internal state of panels while dragging, and syncs it with the
 * layout state in the global manager store when the user is done dragging.
 * Also syncs the layout state from the global manager store to the internal state
 * here when necessary
 */
const useLayoutSyncingState = (
  managerLayoutState: Props['managerLayoutState'],
  setManagerLayoutState: Props['setManagerLayoutState']
) => {
  // ref to keep track of previous managerLayoutState, to check if the props change
  const prevManagerLayoutStateRef = React.useRef<ManagerLayoutState>(managerLayoutState);

  const [internalDraggingSizeState, setInternalDraggingSizeState] = useState<LayoutState>({
    ...managerLayoutState,
    isDragging: false,
  });

  /**
   * Sync FROM managerLayoutState to internalDraggingState if user is not dragging
   */
  useEffect(() => {
    if (
      internalDraggingSizeState.isDragging || // don't interrupt user's drag
      layoutStateIsEqual(managerLayoutState, prevManagerLayoutStateRef.current) // don't set any state if managerLayoutState hasn't changed
    ) {
      return;
    }
    prevManagerLayoutStateRef.current = managerLayoutState;
    setInternalDraggingSizeState((state) => ({ ...state, ...managerLayoutState }));
  }, [internalDraggingSizeState.isDragging, managerLayoutState, setInternalDraggingSizeState]);

  /**
   * Sync size changes TO managerLayoutState when drag is done
   */
  useLayoutEffect(() => {
    if (
      internalDraggingSizeState.isDragging || // wait with syncing managerLayoutState until user is done dragging
      layoutStateIsEqual(prevManagerLayoutStateRef.current, internalDraggingSizeState) // don't sync managerLayoutState if it doesn't differ from internalDraggingSizeState
    ) {
      return;
    }
    const nextState = {
      navSize: internalDraggingSizeState.navSize,
      bottomPanelHeight: internalDraggingSizeState.bottomPanelHeight,
      rightPanelWidth: internalDraggingSizeState.rightPanelWidth,
    };
    prevManagerLayoutStateRef.current = {
      ...prevManagerLayoutStateRef.current,
      ...nextState,
    };
    setManagerLayoutState(nextState);
  }, [internalDraggingSizeState, setManagerLayoutState]);

  const isPagesShown =
    managerLayoutState.viewMode !== 'story' && managerLayoutState.viewMode !== 'docs';
  const isPanelShown = managerLayoutState.viewMode === 'story';

  const { panelResizerRef, sidebarResizerRef } = useDragging(
    setInternalDraggingSizeState,
    isPanelShown
  );
  const { navSize, rightPanelWidth, bottomPanelHeight } = internalDraggingSizeState.isDragging
    ? internalDraggingSizeState
    : managerLayoutState;

  return {
    navSize,
    rightPanelWidth,
    bottomPanelHeight,
    panelPosition: managerLayoutState.panelPosition,
    panelResizerRef,
    sidebarResizerRef,
    showPages: isPagesShown,
    showPanel: isPanelShown,
  };
};

export const Layout = ({ managerLayoutState, setManagerLayoutState, ...slots }: Props) => {
  const {
    navSize,
    rightPanelWidth,
    bottomPanelHeight,
    panelPosition,
    panelResizerRef,
    sidebarResizerRef,
    showPages,
    showPanel,
  } = useLayoutSyncingState(managerLayoutState, setManagerLayoutState);

  return (
    <LayoutContainer
      navSize={navSize}
      rightPanelWidth={rightPanelWidth}
      bottomPanelHeight={bottomPanelHeight}
      panelPosition={managerLayoutState.panelPosition}
      viewMode={managerLayoutState.viewMode}
    >
      {showPages && <PagesContainer>{slots.slotPages}</PagesContainer>}
      <ContentContainer>{slots.slotMain}</ContentContainer>
      <SidebarContainer>
        <Drag ref={sidebarResizerRef} />
        {slots.slotSidebar}
      </SidebarContainer>
      {showPanel && (
        <PanelContainer position={panelPosition}>
          <Drag
            orientation={panelPosition === 'bottom' ? 'horizontal' : 'vertical'}
            position={panelPosition === 'bottom' ? 'left' : 'right'}
            ref={panelResizerRef}
          />
          {slots.slotPanel}
        </PanelContainer>
      )}
    </LayoutContainer>
  );
};

const LayoutContainer = styled.div<ManagerLayoutState>(
  ({ navSize, rightPanelWidth, bottomPanelHeight, viewMode, panelPosition }) => {
    return {
      width: '100%',
      height: '100%',
      display: 'grid',
      overflow: 'hidden',
      gap: 0,
      gridTemplateColumns: `minmax(0, ${navSize}px) minmax(${MINIMUM_CONTENT_WIDTH_PX}px, 1fr) minmax(0, ${rightPanelWidth}px) [right]`,
      gridTemplateRows: `[top] 1fr ${bottomPanelHeight}px [bottom]`,
      gridTemplateAreas: (() => {
        if (viewMode === 'docs') {
          // remove panel in docs viewMode
          return `"sidebar content content"
                  "sidebar content content"`;
        }
        if (panelPosition === 'right') {
          return `"sidebar content panel"
                  "sidebar content panel"`;
        }
        return `"sidebar content content content"
                "sidebar panel   panel   panel"`;
      })(),
    };
  }
);

const SidebarContainer = styled.div(({ theme }) => ({
  backgroundColor: theme.background.app,
  gridArea: 'sidebar',
  position: 'relative',
  borderRight: `1px solid ${theme.color.border}`,
}));

const ContentContainer = styled.div(({ theme }) => ({
  display: 'grid',
  position: 'relative',
  backgroundColor: theme.background.content,
  gridArea: 'content',
}));

const PagesContainer = styled.div(({ theme }) => ({
  // see named areas and tracks in the grid template area defined in LayoutContainer
  gridArea: 'content-start / content-start / right / bottom',
  backgroundColor: theme.background.content,
  zIndex: 1,
}));

const PanelContainer = styled.div<{ position: LayoutState['panelPosition'] }>(
  ({ theme, position }) => ({
    gridArea: 'panel',
    position: 'relative',
    backgroundColor: theme.background.content,
    borderTop: position === 'bottom' ? `1px solid ${theme.color.border}` : null,
    borderLeft: position === 'right' ? `1px solid ${theme.color.border}` : null,
  })
);

const Drag = styled.div<{ orientation?: 'horizontal' | 'vertical'; position?: 'left' | 'right' }>(
  ({ theme }) => ({
    position: 'absolute',
    opacity: 0,
    transition: 'opacity 0.2s ease-in-out',
    zIndex: 2,

    '&:after': {
      content: '""',
      display: 'block',
      backgroundColor: theme.color.secondary,
    },

    '&:hover': {
      opacity: 1,
    },
  }),
  ({ orientation = 'vertical', position = 'left' }) => {
    if (orientation === 'vertical')
      return {
        width: '20px',
        height: '100%',
        top: 0,
        right: position === 'left' ? '-10px' : null,
        left: position === 'right' ? '-10px' : null,

        '&:after': {
          width: 1,
          height: '100%',
          marginLeft: 10,
        },

        '&:hover': {
          cursor: 'col-resize',
        },
      };
    return {
      width: '100%',
      height: '20px',
      top: '-10px',
      left: 0,

      '&:after': {
        width: '100%',
        height: 1,
        marginTop: 9,
      },

      '&:hover': {
        cursor: 'row-resize',
      },
    };
  }
);
