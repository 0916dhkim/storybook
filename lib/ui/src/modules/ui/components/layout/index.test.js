import React from 'react';
import { shallow } from 'enzyme';
import Layout from './index';

jest.mock('react-split-pane', () => function SplitPane() {});

describe('manager.ui.components.layout.index', () => {
  describe('with default options', () => {
    test('should render provided components', () => {
      const wrap = shallow(
        <Layout
          showLeftPanel
          showDownPanel
          goFullScreen={false}
          downPanelInRight={false}
          leftPanel={() => 'LeftPanel'}
          downPanel={() => 'DownPanel'}
          preview={() => 'Preview'}
        />
      );

      expect(wrap).toMatchSnapshot();
    });
  });

  describe('with goFullScreen=true', () => {
    test('should render preview in fullscreen mode', () => {
      const wrap = shallow(
        <Layout
          goFullScreen
          showDownPanel={false}
          showLeftPanel={false}
          downPanelInRight={false}
          leftPanel={() => 'LeftPanel'}
          downPanel={() => 'DownPanel'}
          preview={() => 'Preview'}
        />
      );

      expect(wrap).toMatchSnapshot();
    });
  });

  describe('with showLeftPanel=false', () => {
    test('should hide the leftPanel', () => {
      const wrap = shallow(
        <Layout
          showLeftPanel={false}
          showDownPanel
          downPanelInRight={false}
          goFullScreen={false}
          leftPanel={() => 'LeftPanel'}
          downPanel={() => 'DownPanel'}
          preview={() => 'Preview'}
        />
      );

      expect(wrap).toMatchSnapshot();
    });
  });

  describe('with showDownPanel=false', () => {
    test('should hide the downPanel', () => {
      const wrap = shallow(
        <Layout
          showLeftPanel
          showDownPanel={false}
          goFullScreen={false}
          downPanelInRight={false}
          leftPanel={() => 'LeftPanel'}
          downPanel={() => 'DownPanel'}
          preview={() => 'Preview'}
        />
      );

      expect(wrap).toMatchSnapshot();
    });
  });
});
