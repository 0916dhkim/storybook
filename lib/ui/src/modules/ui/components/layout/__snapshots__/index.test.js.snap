// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager.ui.components.layout.index with default options should render provided components 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#F7F7F7",
      "height": "100vh",
    }
  }
>
  <SplitPane
    allowResize={true}
    defaultSize={250}
    maxSize={-400}
    minSize={150}
    onChange={[Function]}
    onDragFinished={[Function]}
    onDragStarted={[Function]}
    pane1ClassName=""
    pane2ClassName=""
    paneClassName=""
    prefixer={
      Prefixer {
        "_browserInfo": Object {
          "blink": true,
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "name": "Safari",
          "osVersion": NaN,
          "safari": true,
          "x": true,
        },
        "_hasPropsRequiringPrefix": false,
        "_keepUnprefixed": false,
        "_metaData": Object {
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "keepUnprefixed": false,
          "requiresPrefix": Object {},
        },
        "_requiresPrefix": Object {},
        "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
        "prefixedKeyframes": "keyframes",
      }
    }
    primary="first"
    resizerStyle={
      Object {
        "cursor": "col-resize",
        "height": "auto",
        "width": 10,
        "zIndex": 1,
      }
    }
    size={250}
    split="vertical"
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
          "paddingRight": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "flexGrow": 1,
            "height": "100%",
            "width": "100%",
          }
        }
      >
        LeftPanel
      </div>
      <USplit
        shift={5}
        split="vertical"
      />
    </div>
    <SplitPane
      allowResize={true}
      defaultSize={200}
      maxSize={-200}
      minSize={100}
      onChange={[Function]}
      onDragFinished={[Function]}
      onDragStarted={[Function]}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "blink": true,
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": NaN,
            "safari": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": false,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {},
          },
          "_requiresPrefix": Object {},
          "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
          "prefixedKeyframes": "keyframes",
        }
      }
      primary="second"
      resizerStyle={
        Object {
          "cursor": "row-resize",
          "height": 10,
          "width": "100%",
          "zIndex": 1,
        }
      }
      size={200}
      split="horizontal"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "height": "100%",
            "padding": "10px 10px 2px 0",
            "paddingTop": 10,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFF",
              "border": "1px solid #ECECEC",
              "borderRadius": 4,
              "height": "100%",
              "width": "100%",
            }
          }
        >
          Preview
        </div>
        <Dimensions
          height={0}
          width={0}
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "stretch",
            "boxSizing": "border-box",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "padding": "0px 10px 10px 0",
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <USplit
          shift={-5}
          split="horizontal"
        />
        DownPanel
      </div>
    </SplitPane>
  </SplitPane>
</div>
`;

exports[`manager.ui.components.layout.index with goFullScreen=true should render preview in fullscreen mode 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#F7F7F7",
      "height": "100vh",
    }
  }
>
  <SplitPane
    allowResize={false}
    defaultSize={250}
    maxSize={-400}
    minSize={150}
    onChange={[Function]}
    onDragFinished={[Function]}
    onDragStarted={[Function]}
    pane1ClassName=""
    pane2ClassName=""
    paneClassName=""
    prefixer={
      Prefixer {
        "_browserInfo": Object {
          "blink": true,
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "name": "Safari",
          "osVersion": NaN,
          "safari": true,
          "x": true,
        },
        "_hasPropsRequiringPrefix": false,
        "_keepUnprefixed": false,
        "_metaData": Object {
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "keepUnprefixed": false,
          "requiresPrefix": Object {},
        },
        "_requiresPrefix": Object {},
        "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
        "prefixedKeyframes": "keyframes",
      }
    }
    primary="first"
    resizerStyle={
      Object {
        "cursor": undefined,
        "height": "auto",
        "width": 10,
        "zIndex": 1,
      }
    }
    size={1}
    split="vertical"
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "none",
          "flexDirection": "row",
          "paddingRight": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "flexGrow": 1,
            "height": "100%",
            "width": "100%",
          }
        }
      >
        LeftPanel
      </div>
      <USplit
        shift={5}
        split="vertical"
      />
    </div>
    <SplitPane
      allowResize={false}
      defaultSize={200}
      maxSize={-200}
      minSize={100}
      onChange={[Function]}
      onDragFinished={[Function]}
      onDragStarted={[Function]}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "blink": true,
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": NaN,
            "safari": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": false,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {},
          },
          "_requiresPrefix": Object {},
          "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
          "prefixedKeyframes": "keyframes",
        }
      }
      primary="second"
      resizerStyle={
        Object {
          "cursor": undefined,
          "height": 10,
          "width": "100%",
          "zIndex": 1,
        }
      }
      size={1}
      split="horizontal"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "height": "100%",
            "padding": "10px 10px 2px 0",
            "paddingTop": 10,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFF",
              "border": 0,
              "height": "100%",
              "left": "0px",
              "margin": 0,
              "overflow": "auto",
              "padding": 0,
              "position": "fixed",
              "right": "0px",
              "top": "0px",
              "webkitOverflowScrolling": "touch",
              "width": "100%",
              "zIndex": 1,
            }
          }
        >
          Preview
        </div>
        <Dimensions
          height={0}
          width={0}
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "stretch",
            "boxSizing": "border-box",
            "display": "none",
            "flexDirection": "column",
            "height": "100%",
            "padding": "0px 10px 10px 0",
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <USplit
          shift={-5}
          split="horizontal"
        />
        DownPanel
      </div>
    </SplitPane>
  </SplitPane>
</div>
`;

exports[`manager.ui.components.layout.index with showDownPanel=false should hide the downPanel 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#F7F7F7",
      "height": "100vh",
    }
  }
>
  <SplitPane
    allowResize={true}
    defaultSize={250}
    maxSize={-400}
    minSize={150}
    onChange={[Function]}
    onDragFinished={[Function]}
    onDragStarted={[Function]}
    pane1ClassName=""
    pane2ClassName=""
    paneClassName=""
    prefixer={
      Prefixer {
        "_browserInfo": Object {
          "blink": true,
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "name": "Safari",
          "osVersion": NaN,
          "safari": true,
          "x": true,
        },
        "_hasPropsRequiringPrefix": false,
        "_keepUnprefixed": false,
        "_metaData": Object {
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "keepUnprefixed": false,
          "requiresPrefix": Object {},
        },
        "_requiresPrefix": Object {},
        "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
        "prefixedKeyframes": "keyframes",
      }
    }
    primary="first"
    resizerStyle={
      Object {
        "cursor": "col-resize",
        "height": "auto",
        "width": 10,
        "zIndex": 1,
      }
    }
    size={250}
    split="vertical"
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "flex",
          "flexDirection": "row",
          "paddingRight": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "flexGrow": 1,
            "height": "100%",
            "width": "100%",
          }
        }
      >
        LeftPanel
      </div>
      <USplit
        shift={5}
        split="vertical"
      />
    </div>
    <SplitPane
      allowResize={false}
      defaultSize={200}
      maxSize={-200}
      minSize={100}
      onChange={[Function]}
      onDragFinished={[Function]}
      onDragStarted={[Function]}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "blink": true,
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": NaN,
            "safari": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": false,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {},
          },
          "_requiresPrefix": Object {},
          "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
          "prefixedKeyframes": "keyframes",
        }
      }
      primary="second"
      resizerStyle={
        Object {
          "cursor": undefined,
          "height": 10,
          "width": "100%",
          "zIndex": 1,
        }
      }
      size={1}
      split="horizontal"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "height": "100%",
            "padding": "10px 10px 2px 0",
            "paddingTop": 10,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFF",
              "border": "1px solid #ECECEC",
              "borderRadius": 4,
              "height": "100%",
              "width": "100%",
            }
          }
        >
          Preview
        </div>
        <Dimensions
          height={0}
          width={0}
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "stretch",
            "boxSizing": "border-box",
            "display": "none",
            "flexDirection": "column",
            "height": "100%",
            "padding": "0px 10px 10px 0",
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <USplit
          shift={-5}
          split="horizontal"
        />
        DownPanel
      </div>
    </SplitPane>
  </SplitPane>
</div>
`;

exports[`manager.ui.components.layout.index with showLeftPanel=false should hide the leftPanel 1`] = `
<div
  style={
    Object {
      "backgroundColor": "#F7F7F7",
      "height": "100vh",
    }
  }
>
  <SplitPane
    allowResize={false}
    defaultSize={250}
    maxSize={-400}
    minSize={150}
    onChange={[Function]}
    onDragFinished={[Function]}
    onDragStarted={[Function]}
    pane1ClassName=""
    pane2ClassName=""
    paneClassName=""
    prefixer={
      Prefixer {
        "_browserInfo": Object {
          "blink": true,
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "name": "Safari",
          "osVersion": NaN,
          "safari": true,
          "x": true,
        },
        "_hasPropsRequiringPrefix": false,
        "_keepUnprefixed": false,
        "_metaData": Object {
          "browserName": "safari",
          "browserVersion": NaN,
          "cssPrefix": "-webkit-",
          "jsPrefix": "Webkit",
          "keepUnprefixed": false,
          "requiresPrefix": Object {},
        },
        "_requiresPrefix": Object {},
        "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
        "prefixedKeyframes": "keyframes",
      }
    }
    primary="first"
    resizerStyle={
      Object {
        "cursor": undefined,
        "height": "auto",
        "width": 10,
        "zIndex": 1,
      }
    }
    size={1}
    split="vertical"
  >
    <div
      style={
        Object {
          "alignItems": "stretch",
          "display": "none",
          "flexDirection": "row",
          "paddingRight": 0,
          "width": "100%",
        }
      }
    >
      <div
        style={
          Object {
            "flexGrow": 1,
            "height": "100%",
            "width": "100%",
          }
        }
      >
        LeftPanel
      </div>
      <USplit
        shift={5}
        split="vertical"
      />
    </div>
    <SplitPane
      allowResize={true}
      defaultSize={200}
      maxSize={-200}
      minSize={100}
      onChange={[Function]}
      onDragFinished={[Function]}
      onDragStarted={[Function]}
      pane1ClassName=""
      pane2ClassName=""
      paneClassName=""
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "blink": true,
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": NaN,
            "safari": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": false,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": NaN,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {},
          },
          "_requiresPrefix": Object {},
          "_userAgent": "Node.js (darwin; U; rv:v8.7.0) AppleWebKit/537.36 (KHTML, like Gecko)",
          "prefixedKeyframes": "keyframes",
        }
      }
      primary="second"
      resizerStyle={
        Object {
          "cursor": "row-resize",
          "height": 10,
          "width": "100%",
          "zIndex": 1,
        }
      }
      size={200}
      split="horizontal"
    >
      <div
        style={
          Object {
            "boxSizing": "border-box",
            "height": "100%",
            "padding": "10px 10px 2px 0",
            "paddingTop": 10,
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <div
          style={
            Object {
              "backgroundColor": "#FFF",
              "border": "1px solid #ECECEC",
              "borderRadius": 4,
              "height": "100%",
              "width": "100%",
            }
          }
        >
          Preview
        </div>
        <Dimensions
          height={0}
          width={0}
        />
      </div>
      <div
        style={
          Object {
            "alignItems": "stretch",
            "boxSizing": "border-box",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "padding": "0px 10px 10px 0",
            "position": "absolute",
            "width": "100%",
          }
        }
      >
        <USplit
          shift={-5}
          split="horizontal"
        />
        DownPanel
      </div>
    </SplitPane>
  </SplitPane>
</div>
`;
